feat: Editorial review: Speculation rules Chrome 121 (#32455)
mdn/content: files/en-us/web/html/element/script/type/speculationrules/index.md

Speculation rules take the form of a JSON structure that determine what resources should be prefetched or prerendered by the browser. This is part of the {{domxref("Speculation Rules API", "", "", "nocode")}}.

> > **Note:** As an alternative, speculation rules can be defined inside external text files referenced by the {{httpheader("Speculation-Rules")}} HTTP header. Specifying an HTTP header is useful in cases where developers are not able to directly modify the document itself.

Reviewer: Perhaps ... (also linking to the "JSON representation provided below).
```suggestion
> **Note:** Speculation rules can be defined inside external text files referenced by the {{httpheader("Speculation-Rules")}} HTTP header, using the same JSON representation provided below. Specifying an HTTP header is useful in cases where developers are not able to directly modify the document itself.
```

Author: Like it. Updated, and link added.

- `"urls"`

  - : An array of strings representing a list of URLs to apply the rule to. These can be absolute or relative URLs. Relative URLs will be parsed relative to the document base URL (if inline in a document) or relative to the external resource URL (if externally fetched). `"urls"` and `"where"` cannot both be set in the same rule.

- `"where"`

  - : An object representing the conditions by which the rule matches URLs contained in the associated document. Effectively, the `"where"` object represents a test that is performed on every link on the page to see whether the speculation rule is applied to it. `"where"` and `"urls"` cannot both be set in the same rule.

    This object can contain exactly one of the following properties:

    - `"href_matches"`
      - : A string containing a URL pattern, or an array containing multiple URL pattern strings, which follow the standard [URL Pattern API syntax](/en-US/docs/Web/API/URL_Pattern_API). Links in the document whose URLs match the pattern(s) will have the rule applied.
    - `"relative_to"`

>       - : In the case of an `"href_matches"` condition, this can specify where you want that condition to be matched relative to. This works in exactly the same way as the [rule-level `"relative_to"` key](/en-US/docs/Web/HTML/Element/script/type/speculationrules#relative_to), except that it only affects a single `"href_matches"` condition inside a `"where"` key.

Reviewer: This is now definitely a circular link :-). This should fix it. FWIW I think this does make it clear they are different.

```suggestion
      - : In the case of an `"href_matches"` condition, this can specify where you want that condition to be matched relative to. This works in exactly the same way as the [rule-level `"relative_to"` key](/en-US/docs/Web/HTML/Element/script/type/speculationrules##relative_to_2), except that it only affects a single `"href_matches"` condition inside a `"where"` key.
```

Author: Ack, dammnit! Fixed now ;-)

> Specifically, each object can contain the following properties:

Reviewer: ```suggestion
Each object can contain the following properties:
```

Author: updated

- `"source"`

>   - : A string representing the URLs to which the rule applies. This can be one of:

Reviewer: It looks like this never needs to be specified right - i.e. it can always be inferred. Perhaps we should say that here. Something like ...

```suggestion
  - : A string indicating which other properties are used to specify the set of prefetched/rendered URLs. This is optional because the value can always be inferred from the other properties.
     This can be one of:
```

Author: like it; updated

Reviewer 2: Note this inference is new in Chrome 122. Prior to that you had to be explicit:
https://developer.chrome.com/blog/speculation-rules-improvements#optional_source

Reviewer: @chrisdavidmills Based on what @tunetheweb said, I will assume you have got a BCD reference for "source_key_optional"? Of course in 30 releases no one will be using anything before 122 and it won't matter.

Author: I have just filed one: https://github.com/mdn/browser-compat-data/pull/22701

    - `"document"`

>       - : Specifies that the URLs will be matched from links contained in the associated document (i.e. pointed to by {{htmlelement("a")}} and {{htmlelement("area")}} elements), based on the conditions described by a `"where"` key. Note that the presence of a `"where"` key implies `"source": "document"`, so it is optional.

Reviewer: ```suggestion
      - : Specifies that the URLs will be matched from links in the associated document (as defined in {{htmlelement("a")}} and {{htmlelement("area")}} elements), based on the conditions described by a `"where"` key. Note that the presence of a `"where"` key implies `"source": "document"`, so it is optional.
```

Author: updated

    - `"list"`
      - : Specifies that the URLs will come from a list, specified in the `"urls"` key. Note that the presence of a `"urls"` key implies `"source": "list"`, so it is optional.

- `"urls"`

>   - : An array of strings representing a list of URLs to apply the rule to, in the case of a `"source": "list"` rule. These can be absolute or relative URLs. Relative URLs will be parsed relative to the document base URL (if inline in a document) or relative to the external resource URL (if externally fetched).

Reviewer: ", in the case of a `"source": "list"` rule." feels a bit odd/unnecessary.

Either way though, I the important thing here is not that it is a list rule, but that you can't also use `where` if you set this.

Author: I've removed those lines from `urls` and `where`, and added a line to each saying that both cannot be set in the same rule.

- `"where"`

>   - : An object representing the conditions by which the rule matches URLs contained in the associated document, in the case of a `"source": "document"` rule. Effectively, the `"where"` object represents a test that is performed on every link on the page to see whether the speculation rule is applied to it. This object can contain exactly one of the following properties:

Reviewer: As above ", in the case of a `"source": "document"` rule" reads oddly.

Reviewer: But I like this line "Effectively, the `"where"` object represents a test that is performed on every link on the page to see whether the speculation rule is applied to it."

Author: See above comment.

I think "navigation link" is somewhere redundant, yes.

>     - `"href_matches"`: A string containing a URL pattern, or an array containing multiple URL pattern strings, which follow the standard [URL Pattern API syntax](/en-US/docs/Web/API/URL_Pattern_API). Links in the document whose URLs match the pattern(s) will have the rule applied.

Reviewer: I'd highly recommend making these definition lists. Always nice to have somewhere to link to.

Author: Done. Nested `<dl>`s are a lot more usable now that the left-hand line style has been added.

>     - `"relative_to"`: In the case of an `"href_matches"` condition, this can specify where you want that condition to be matched relative to. This works in exactly the same way as the rule-level [`"relative_to"`](/en-US/docs/Web/HTML/Element/script/type/speculationrules#relative_to) key, except that it only affects a single `"href_matches"` condition inside a `"where"` key.

Reviewer: This confuses me, it seems to link back to itself, and it sounds like there is another relative-to definition I can't see.

Reviewer: Ah, it is further down.

Author: I've expanded the link to "rule-level `relative_to` key", to provide a stronger hint that this is something different.

Reviewer: The approach is good, but I think the link might be wrong

    - `"not"`: An object containing one condition (`"href_matches"`, `"selector_matches"`, `"and"`, `"not"`, or `"or"`) which, if it matches, will _not_ have the rule applied to it. All links that _do not_ match the condition _will_ have the rule applied.
    - `"or"`: An array containing one or more objects containing conditions (`"href_matches"`, `"selector_matches"`, `"and"`, `"not"`, or `"or"`), any of which can match for the rule to be applied to them.

    `"where"` conditions can be nested multiple levels deep to create complex conditions, or you can choose to split them into separate rules to keep them simple. See [where examples](#where_examples) for more explanation, and multiple examples of use.

- `"eagerness"`

  - : A string providing a hint to the browser as to how eagerly it should prefetch/prerender link targets in order to balance performance advantages against resource overheads. Possible values are:

    - `"immediate"`: The author thinks the link is very likely to be followed, and/or the document may take significant time to fetch. Prefetch/prerender should start as soon as possible, subject only to considerations such as user preferences and resource limits.

>     - `"eager"`: The author wants to prefetch/prerender a large number of navigations, as early as possible. Prefetch/prerender should start on any slight suggestion that a link may be followed. For example, the user could move their mouse cursor towards the link, hover/focus it for a moment, or pause scrolling with the link in a prominent place.

Reviewer: Indicate that this is the default

    - `"moderate"`: The author is looking for a balance between `eager` and `conservative`. Prefetch/prerender should start when there is a reasonable suggestion that the user will follow a link in the near future. For example, the user could scroll a link into the viewport and hover/focus it for some time.
    - `"conservative"`: The author wishes to get some benefit from speculative loading with a fairly small tradeoff of resources. Prefetch/prerender should start only when the user is starting to click on the link, for example on {{domxref("Element.mousedown_event", "mousedown")}} or {{domxref("Element.pointerdown_event", "pointerdown")}}.

>     If not specified, `"source": "list"` rules default to `immediate` and `"source": "document"` rules default to `conservative`. The browser takes this hint into consideration along with its own heuristics, so it may select a link that the author has hinted as less eager than another, if the less eager candidate is considered a better choice.

Reviewer: I wonder if we should specify in terms of `urls` and `where` rules? The source is kind of irrelevant if it can be inferred.

Reviewer 2: Agreed. I would either do it in terms of `urls` or `where` or say "list rules" or "document rules" rather than `"source": "list"` and `"source": "document"`.

Author: I've changed it to "...list (`"urls"`) rules default to `immediate` and document (`"where"`) rules..."

they are commonly called list rules and document rules.

- `"expects_no_vary_search"`

>   - : A string providing a hint to the browser as to what the expected {{httpheader("No-Vary-Search")}} header value will be (if any) for documents that it is receiving prefetch/prerender requests for via the speculation rules. The browser can use this to determine ahead of time whether it is more useful to wait for an existing prefetch/prerender to finish, or start a new fetch request when the speculation rule is matched. See the [`"expects_no_vary_search"` example](#expects_no_vary_search_example) for more explanation of how this can be used.

Reviewer: That first sentence is a mouthful. I don't really get this feature, or perhaps what the name of it implies. 

I mean I don't understand why you would not wait for an existing prerender to finish - what does getting back a mismatch in this header mean? Anyway, perhaps this is answered in the example.

Author: This is a bit mindblowing, and took me a while to understand. Basically, `No-Vary-Search` defines rules that determine whether similar URLs (i.e. same path, but different params, or same params in different order) should be considered identical for caching purposes. This is useful because, for example, if `/users` and `/users?id=123` display exactly the same page but with a bit of client-side logic to highlight a specific user name, there is no point wasting time caching another copy of the same page — just use the page already in the cache.

However, for speculation rule purposes, it is useful to know ahead of time what this behavior is intended to be. Is it useful to wait for a prerender of `/users` to finish when you are on that page, but have clicked a link to `/users?id=123`? Or is it a waste of time, because the pages are actually different, so a new pretender will be needed anyway?

I'd updated the first sentence to try to make it less confusing.

    > **Note:** A cross-site prefetch requires a referrer policy that is at least as strict as the default `"strict-origin-when-cross-origin"` value — so `"strict-origin-when-cross-origin"`, `"same-origin"`, `"strict-origin"`, or `"no-referrer"`. A laxer policy set in the speculation rules will override a stricter policy set on the referring page as long as it is still sufficiently strict for the cross-site case.

    > **Note:** In the case of document rules, the matched link's specified referrer policy (e.g. using the [`referrerpolicy`](/en-US/docs/Web/HTML/Element/a#referrerpolicy) attribute) will be used, unless the rule specifies a policy that overrides it.

- `"relative_to"`

  - : A string specifying where you want links matched by URL to be matched relative to. The value can be one of:

    - `document`: URLs should be matched relative to the document the speculation rules are being set on.
    - `ruleset`: URLs should be matched relative to the file the rules are specified in. This is the default value.

>     When rules are specified inside the same document they are being set for (i.e. in an inline `<script>` element), this key setting makes no difference. It is significant when the rules are contained in an external file (i.e. as is the case when using the {{httpheader("Speculation-Rules")}} header to specify rules).

Reviewer: I would invert

```suggestion
    This setting is only relevant for rules defined in an external file (set using the {{httpheader("Speculation-Rules")}} header). When rules are specified inside the same document they are being set for (i.e. in an inline `<script>` element), this key setting makes no difference..
```

Author: done

- `"requires"`

  - : An array of strings representing capabilities of the browser parsing the rule, which must be available if the rule is to be applied to the specified URLs.

> 

Author: OK; I've updated it to make it non-code.

        - A future Safari implementation may possibly use something along the lines of [iCloud Private Relay](https://support.apple.com/en-us/102602).
        - A future Firefox implementation might use something based on the [Mozilla VPN](https://www.mozilla.org/en-US/products/vpn/) product.

> **Note:** As speculation rules use a `<script>` element, they need to be explicitly allowed in the [`Content-Security-Policy`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) [`script-src`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src) directive if the site includes it. This is done by adding the `"inline-speculation-rules"` value along with a hash- or nonce-source.

## Examples

### Prefetch and prerender in the same set of rules

The basic examples shown in the description section included separate speculation rules defined for prefetch and prerender. It is possible to define both in a single set of rules:

```html
<script type="speculationrules">
  {
    "prefetch": [

>       {

Reviewer: Can we use `which` and `urls` for prefetch and prerender - show an example and note that?

Author: done

        "urls": ["next.html", "next2.html"],
        "requires": ["anonymous-client-ip-when-cross-origin"],
        "referrer_policy": "no-referrer"
      }
    ],
    "prerender": [
      {
        "urls": ["next3.html", "next4.html"]
      }
    ]
  }
</script>
```

### Multiple rule sets

It is also allowable to include multiple sets of rules in a single HTML file:

> 

Reviewer: Below, at least some of these should use `which`. Mosty because that is likely to be preferred.

Author: I've done the first ask here. 

I don't want to add a separate section for `list` examples. The point of the `where` examples section is to cover all the different syntaxes that can be used in `where`. `list` syntax is really simple in comparison.

Reviewer: The important thing for me is that `where` is the prominent and preferred option - my assumption is that is the intent, since to my mind it simplifies things quite a lot.

So perhaps below and in any example here, switch the where example before the list example?

Author: I think it is likely to be more common, but at the same time, both are valid for different use cases, and I don't think it is MDN's place to favor one over the other. I also don't think the effort involved in doing this is worth it. No change.

Reviewer 2: IMHO `where` is likely to be used in a single ruleset.
`url` is likely to be used where the page wants more control so will be inserting rulessets of different urls as and when it sees fit.

So I agree with @chrisdavidmills that we should leave examples for `url`.

Reviewer: Sounds reasonable!

You can see this in action in this [prerender demos](https://prerender-demos.glitch.me/) page.

### `where` examples

> A document-sourced rule contains a `"where"` property, which is equal to an object containing criteria that define which links in the document are matched. Effectively, the `"where"` object represents a test that is performed on every link on the page to see whether the speculation rule is applied to it.

Reviewer: ```suggestion
A document-sourced rule contains a `"where"` property, which is an object containing criteria that define which links in the document are matched. Effectively, the `"where"` object represents a test that is performed on every link on the page to see whether the speculation rule is applied to it.
```

Author: updated

The most basic version will match a single URL pattern or CSS selector:

```json
{ "where": { "href_matches": "/next" } }
```

```json
{ "where": { "selector_matches": ".important-link" } }
```

> `"href_matches"` and `"selector_matches"` can also be set equal to arrays of values, so multiple URL patterns or CSS selectors can be matched simultaneously:

Reviewer: ```suggestion
`"href_matches"` and `"selector_matches"` can also be set to an array of values, so multiple URL patterns or CSS selectors can be matched simultaneously:
```

Author: updated

```json
{ "where": { "href_matches": ["/next", "/profile"] } }
```

```json
{ "where": { "selector_matches": [".important-link", "#unique-link"] } }
```

> URL patterns and selectors can also contain wildcard (`*`) characters, allowing a single value to match multiple URLs. The following for example could match `user/`, `user/settings`, `user/stats`, etc.

Reviewer: I feel a comma might help.
```suggestion
URL patterns and selectors can also contain wildcard (`*`) characters, allowing a single value to match multiple URLs. For example, the object below could match `user/`, `user/settings`, `user/stats`, etc.
```

Author: updated

#### Combining multiple `"where"` conditions with `"and"` or `"or"`

Multiple conditions can be combined inside `"and"` or `"or"` conditions — these take the value of arrays containing multiple conditions, all or any of which (respectively) have to match for the speculation rules to apply to a link. Using `"and"` or `"or"`, conditions can be nested multiple levels deep — there is no specified limit on allowed nesting levels.

It is useful to think of the `"where"` object as being equivalent to an `if` statement. So

```text
{ and: [A, B, { or: [C, { not: D }] }] }
```

is equivalent to

```text
if (A && B && (C || !D)) {
  apply speculation rule
}
```

In the following complete speculation rule example, all same-origin pages are marked for prefetching except those known to be problematic — the `/logout` page, and any links marked up with a class of `.no-prerender`:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "where": {
          "and": [
            { "href_matches": "/*" },
            { "not": { "href_matches": "/logout" } },
            { "not": { "selector_matches": ".no-prerender" } }
          ]
        }
      }
    ]
  }
</script>
```

> > **Note:** The `where` pattern above excludes cross-site links, which are supported for prefetching (provided the user has no cookies set for the destination site, to protect against tracking) but not for prerendering.

Reviewer: These examples are good. I presume if I didn't exclude cross-site links it wouldn't make any difference.

Author: We are including relative links of any kind, but not ones that match the pattern `/logout` or have a class of `.no-prerender` on them.

If you also included cross-site links, e.g. with `{ "href_matches": "/*" }`, you'd also prerender all cross-site links, except ones that contain `/logout` or have a class of `.no-prerender` on them.

Reviewer 2: @chrisdavidmills is that last comment I think you mean without the `/`?:

> If you also included cross-site links, e.g. with { "href_matches": "*" }

And you'd attempt to prefetch cross-site links in that case, but only succeed if they returned that extra header.

Reviewer: OK, so you're not actually excluding cross-site links - you're just no including them right? Should we also mention the extra header? Below highlights what I mean in "pseudo text".

```suggestion
> **Note:** The `where` pattern above does not include cross-site links, which are supported for prefetching (provided the user has no cookies set for the destination site, to protect against tracking, and the site returns the header XXXX?), but not for prerendering.
```

Author: @tunetheweb yes, sorry, I meant `"*"`! I'm not sure which extra header you mean - can you elaborate? The one that springs to mind is `Supports-Loading-Mode`, but I thought the opt-in for cross-site prefetch wasn't supported yet, just the opt-in for cross-origin, same-site prerender?

Reviewer 2: Ah good point. Didn't realise this comment was about prefetch. Ignore me.

Author: No worries, glad we're cleared that up.

### `"expects_no_vary_search"` example

Consider the case of a user directory landing page, `/users`, which has an `id` parameter added to bring up information on a specific user, for example `/users?id=345`. Whether this URL should be considered identical for caching purposes depends on the behavior of the application:

1. If this parameter has the effect of loading a completely new page containing the information for the specified user, then the URL should be cached separately.
2. If this parameter has the effect of highlighting the specified user on the same page, and perhaps revealing a pullout panel displaying their data, then the URL should be considered the same for caching purposes. This could result in performance improvements around the loading of the user pages and could be achieved via a {{httpheader("No-Vary-Search")}} with a value of `params=("id")`.

How does this affect speculation rules? Consider the following code:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "urls": ["/users"]
      }
    ]
  }
</script>
<a href="/users?id=345">User Bob</a>
```

> What would happen in this case when the user starts a navigation to `/users?id=345` when the headers for the prefetch of `/users` have not been received yet? At this point, the browser doesn't know what the `No-Vary-Search` value will be, if anything. If there was no `No-Vary-Search` value set, and the application behavior was more like Option 1 above, the prefetch would be wasted and the browser would need to go and fetch the separate `/users?id=345` page from scratch.

Reviewer: So if the user didn't start a navigation before we prefetched we'd get `No-Vary-Search: params=("id")` with the returned resource. At this point we know that we don't have to download again, because this says any page is OK with that id. Right?

What if it comes back empty or doesn't come back? Presumably that means that every variant needs to be cached, which means that we will have to re-request for the specific URL?

I guess what I'm saying is it would be good to talk about option 1 here too.

Reviewer: Another thing to point out, is that I don't think this matters if we had used `which`. In that case the URL pattern would have matched the link as specified, NOT the URL that was downloaded as per the urls  key - right? So with which you can ignore all this.

Author: Yeah, it doesn't matter whether list or document rules are used, you are saying that you expect the `No-Vary-Search` header to be a certain value for those matched URLs, allowing the browser to make smarter decisions ahead of time.  

If the user didn't start a navigation before the prefetch came back with `No-Vary-Search: params=("id")` set, this would be kind of irrelevant for this page.

I'm not sure anything else needs to be added here.

Reviewer: @chrisdavidmills I understand that if you specify `expects_no_vary_search` the behaviour will be the same whether  list or document rules are used.

That's not my point. I THINK that if you're using `which` it would be pointless to also specify `expects_no_vary_search`, and if I am correct, that is worth saying (i.e. this is only relevant to specific URLs). 

The reason is that if you specify `which` then you're actually prefetching the URL that is in the document. So if they navigate to it early it will be the same thing that is being prefetched (in this case `users.html?id=345`). There's no question about whether you will want to discard it or not - you want to keep it.

I suspect there is no harm in specifying `expects_no_vary_search` but also there is no benefit. 

This section was the hardest thing in the document for me to understand, so if we can tell be people the conditions under which they don't need to worry about it (if there are any) that's a good thing right?

Reviewer 2: With `which` , you could hover over `users.html?id=123`, and before that finishes, hover over `users.html?id=345`.

The `expects_no_vary_search` will allow you to hold off fetching `users.html?id=345` since we know one we expect to match (`id=123`) that is already in flight and that'll do fine.

Author: OK, I think I see what you mean now. If you were using the following in your speculation rule:

```
{ "where": { "href_matches": "/users*" } }
```

Then `expects_no_vary_search` wouldn't make much difference, as you'd be prefetching both `/users` and `/users?id=345` if they were linked in the document?

But I guess it could still make a difference, depending on what pattern you specified to match. For example:

```
{ "where": { "href_matches": "/users" } }
```

Would not match `/users?id=345`, right?

But I'm not sure. @tunetheweb, can you help clear up this confusion?

Author: OK, I think I'm getting there now, thanks for the help! I have added a where example to the bottom of the section. Does this make sense:

Document rules can also be used in conjunction with `"expects_no_vary_search"`, depending on the pattern used. For example, in the case of:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        { "where": { "href_matches": "/users?id=*" } },
        "expects_no_vary_search": "params=(\"id\")"
      }
    ]
  }
</script>
<a href="/users?id=012">User Bill</a>
<a href="/users?id=345">User Bob</a>
<a href="/users?id=678">User Ben</a>
```

A copy of the `/users` page would start to be prefetched when one of the links was hovered over, but only one, as all three would be counted as the same page for caching purposes.

?

Reviewer 2: Yeah that works.

Reviewer: Great to add an example - and @tunetheweb thanks for the example of hover over one link with particular id then hover over another. Makes sense that in this case you also need the hint so you don't see the new ID and decide you have to abort the original prefetch. 

@chrisdavidmills I'll propose just a slight variation on this.

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "urls": ["/users"],
        "expects_no_vary_search": "params=(\"id\")"
      }
    ]
  }
</script>
<a href="/users?id=345">User Bob</a>
```

This indicates that Option 2 described above is what the server is expected to produce. If a navigation starts while there is an ongoing prefetch of `/users`, this informs the browser that it is appropriate to wait for the prefetch, instead of immediately starting another fetch for `/users?id=345`.

Document rules can also be used in conjunction with `"expects_no_vary_search"`, depending on the pattern used. For example, in the case of:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        { "where": { "href_matches": "/users?id=*" } },
        "expects_no_vary_search": "params=(\"id\")"
      }
    ]
  }
</script>
<a href="/users?id=012">User Bill</a>
<a href="/users?id=345">User Bob</a>
<a href="/users?id=678">User Ben</a>
```

> A copy of the `/users` page would start to be prefetched when one of the links was hovered over, but only one, as all three would be counted as the same page for caching purposes.

Reviewer: While it is true that "A copy of the `/users` page will be fetched", I think it is the _specific_ URL you selected. I think this variant possibly requires a little less thought by the user about what is going on:

```suggestion
If a link is hovered over, the browser will start prefetching that specific link.
If the user hovers over another link before the prefetch completes the `expects_no_vary_search` pattern tells the browser that there is no need to cancel the current prefetch, because all id values are effectively the same page for this context (and for caching purposes).
```
